Clarify CRAM section 7.1 BAM header in first container.

cmnbroad opened the pull request:
Proposed clarification for section CRAM spec section 7.1. See issue https://github.com/samtools/hts-specs/issues/450.

jkbonfield commented:
I think we should change the text to SAM header instead of BAM header, or better still to CRAM header with a separate paragraph stating that at present the CRAM header format is aligned to the SAM one.  Both BAM and CRAM are compressed representations of SAM and it's not helpful to couch CRAM in terms of BAM.  Additionally BAM actually does have an additional binary header component, so it's ambiguous.

The header is purely textual and matches the SAM specification (but I wouldn't want to paint ourselves into a corner - we may at some point wish to add a new `@xy` style meta-data item purely for use within CRAM).  

Regarding the padding, I explored several ideas when implementing this.  Prior to CRAM 3.0 I just expanded the block and terminated the string with much more than 1 nul.  As you point out that doesn't work for compressed headers.  Hence when we bumped to version 3 we also permitted more than one block to be present, so we could compress the text one and leave an uncompressed block for expansion purposes.  It looks like I add either 50% or 10K - whichever is smaller - but that's purely an implementation decision.  I don't object to mentioning it in a recommendation though.

I did also explore a third route, which was to increase the container size and simply have unallocated space after the block.  Eg a 10K container that holds a 5K compressed block.  I can't recall whether I ran into problems with it, but at some point I dismissed this scheme.

Agreed we should clarify how we want it to work in cram 3.0, which is simply to say we permit additional blocks which may be used as a way to reserve space for header expansion, permitting efficient header replacement.

PS.  I don't understand why we only permitted gzip for compressing the header.  Why not bzip2 for example?  One possible reason could perhaps be we were considering adding extra meta-data to explain codecs used.  This is sadly lacking.  A program that doesn't have lzma support could be happily decoding GBs of CRAM and then suddenly meet an lzma block.  There's no way for it to know it was permitted and may exist somewhere in the file.

jkbonfield commented on CRAMv3.tex:
```diff
@@ -477,15 +477,17 @@ \section{\textbf{Container header structure}}
+It is recommended to reserve 50\% more space in the BAM header container than
+is required for the BAM header by optionally padding the container with a second
+raw block consisting of all zeroes. This can be used to subsequently expand the header
+container in place, such as when updating @SQ records, while preserving the relative
```
"while preserving the *absolute* offsets of all subsequent containers".

cmnbroad replied:
Oh yeah...Done.

jkbonfield commented on CRAMv3.tex:
```diff
@@ -477,15 +477,17 @@ \section{\textbf{Container header structure}}
-records additional space may be required for the BAM header. It is recommended
-to reserve 50\% more space in the CRAM header block than it is required by the
-BAM header.
+The first container in a CRAM file contains the BAM header in an optionally
```
Maybe something like:

> The first container in a CRAM file contains a textual header in a single block, optionally gzip compressed.
> This text header currently matches the SAM header specification.

From that point on we can refer to CRAM header instead of BAM header.  I can replace the images with tikz pic if you wish.

I say "single block" because that's how it is currently defined, but for future versions I want to amend this to "one or more blocks", or to permit blocks to be able to be more than 2GB in size.  I say this because people have already hit limits in both BAM and CRAM with *huge* headers due to millions of SQ lines.  If we keep blocks are 32-bit sizes, then we'd need to permit multiple blocks, with obvious concatenation properties to construct the header.

I guess it's such a trivially backwards compatible route (multi-blocks, not 64-bit sizes) that we could add that to CRAM 3.1 too.  If we chose that route in 3.1 then we'd also amend the next paragraph with "additional raw blocks" instead of "a second raw block".

cmnbroad commented:
I updated this based on comments (I **think** it follows what you suggested), and while I was at it, I added the end-of-file container to the file structure section, since it really should be there. The images do need to be updated (all this latex stuff is pretty new to me so if you think tikz is the way to go that sounds good to me).

jkbonfield commented:
Looks good.  I'll update the pictures and, if I may, push an extra commit onto your branch.

cmnbroad commented:
Thanks- feel free to update.

jkbonfield commented:
See my tikzification commit.

I had to move text around a little and reword some as the figure legends weren't particularly well placed and also seemed to be a bit repetitive, just duplicating the setence before.  Feel free to adjust and tweak as you see fit.

cmnbroad commented:
@jkbonfield I think these changes look pretty good- thanks for the tikz updates. Not sure what you mean by "the last figure hasn't yet been rewritten" ?

jkbonfield commented:
A few pages on there is another figure, in a different style but still a .png.

jkbonfield commented on CRAMv3.tex:
```diff
@@ -964,23 +964,104 @@ \subsection{\textbf{Core data block}}
+\filldraw[draw=black, fill=white!10, drop shadow] (1.5, -6.1) ellipse (1.8 and 0.5);
+\filldraw[draw=black, fill=white!10, drop shadow] (1.5, -7.6) ellipse (1.8 and 0.5);
+
+\node [right=of p1, scale=0.7, text centered, text width=14em] {Core (Bit) Encodings (Huffman, RANS, etc.)};
```
RANS isn't a bit encoding; it only applies to external blocks.  Instead "Huffman, Beta, etc." would be appropriate.

I'm also a little wary of the layout as it sort of implies specifics which the old picture didn't.  We previously had big fat arrows that spanned multiple items on the left and weren't well aligned with them.  The new one looks to have a deliberate one to one alignment with specific types of data.  While this is just an example, I'd be cautious not to hint that specific types are encoding in specific ways.

If  we wish to keep this image, we'll want to change the legend from things like "how a CRAM record _is_ distributed" to "how a CRAM record _may be_ distributed".

PS. More complex Tikz work than I've done so far - a good opener!   One of the reasons I didn't do this was because it'd mean spending more time, which I didn't have, learning Tikz. :-)

jkbonfield replied:
The flip side to the above argument, if we are happy with a diagram showing an example rather than describing the basic concepts without implying anything concrete,  then we should perhaps be more bold and give a realistic example.

Eg using specific types of byte encodings (EXTERNAL,  BYTE_ARRAY_STOP, etc) on obvious numerics and strings.  We could then also add appropriate examples of compression codecs.  Eg BYTE_ARRAY_STOP on read identifiers to an external block compressed using bzip2,  EXTERNAL for quality scores compressed with rANS, and so on.  Maybe pick Bit flags, alignment start for HUFFMAN and BETA encodings.

We need to keep a clear separation of encoding (which may be byte based or bit based) and compression codec (as-is, gzip, bzip2, lzma, rans).  Especially as this is likely something we'll expand on in future.

cmnbroad replied:
Ok, thx for the comments. Point taken on RANS, thats in the wrong place. Also, I had tried to make the left side of the arrows start from the `CRAM Record` box rather than the individual fields, and the right side end at the individual blocks, but perhaps thats too subtle. I'll do another pass.

cmnbroad commented:
@jkbonfield I took another shot at the image - see what you think.

jkbonfield commented:
> @jkbonfield I took another shot at the image - see what you think.

Looks great.  Thanks.  Gtg.
